docs/source/traits_user_manual/observers.rst

```diff
@@ -0,0 +1,557 @@
+   * - |Expression.set_items|
+     - For observing items in a set.
+   * - |Expression.filter_|
+     - For observing traits satisfying a user-defined filter.
+   * - |Expression.parse|
+     - For parsing a string written in Traits domain specific language into an
+       expression object.
+
+Users can combine complex expressions using the functions |Expression.then|,
+|Expression.join_| or Python's bitwise-or (`|`) operation.
+
+Most of these functions will have a ``notify`` argument for setting whether
+notifications should be fired for changes.
+
+.. rubric:: Example expressions
+
+.. list-table::
+   :widths: 15 25
+   :header-rows: 1
+
+   * - Expression
+     - Meaning
+   * - *trait("attr1")*
+     - Matches a trait named *attr1* on an object and notifies for changes.
+   * - *trait("attr1", optional=True)*
+     - Matches a trait named *attr1* if it is defined. Ignore if it is not
+       defined.
+   * - *trait("attr1", notify=False).trait("attr2")*
+     - Matches a trait named *attr2* on an object referenced by a trait named
+       *attr1* on the current object. Changes to *attr2* will trigger
+       notifications, while changes to *attr1* do not.
+   * - *trait("foo").list_items().list_items().trait("value")*
+     - Matches the *value* trait on an item of a nested list in another list
+       *foo*. Assignment changes to *foo*, mutations to the lists or changes
+       to *value* will trigger notifications.
+   * - *metadata("updated")*
+     - Matches any trait on the current that has a metadata attribute named
+       *updated*. Changes on those traits will trigger notifications.
+   * - *trait("foo") | trait("bar")*
+     - Matches trait named *foo* or *bar* on the current object. Changes on
+       *foo* or *bar* will trigger notifications.
+   * - *trait("foo").then(trait("bar") | trait("baz"))*
+     - Matches *foo.bar* or *foo.baz* on the current object. Changes on *foo*,
+       *foo.bar* or *foo.baz* will trigger notifications.
+   * - *trait("foo").filter_(lambda n, t: True)*
+     - Matches any traits on *foo* on the current object. Changes on *foo* or
+       the nested attributes will trigger notifications.
+
+
+.. rubric:: Describe an expression
+
+Expression can be defined as a standalone object without being attached
+to an instance of |HasTraits|. To understand what an expression means, its
+|Expression.print| method can be used::
+
+    >>> from traits.observers.api import trait
+    >>> trait("children").list_items().trait("age").print()
+    Observes trait named 'children' with notifications,
+        then observes items in a list with notifications,
+            then observes trait named 'age' with notifications.
+
+
+Traits domain specific language
+```````````````````````````````
+Traits domain specific language (Traits DSL) provides a convenient and concise
+way to specify observation rules via a single text. It supports a subset of
+features in |Expression|.
+
+.. rubric:: Syntax
```

Comment:
The formal grammar probably doesn't belong here, but it would certainly fit well in an "internals" section, or as part of the main "internals" chapter.